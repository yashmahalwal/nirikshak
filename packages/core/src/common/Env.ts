/* istanbul ignore file */
/* eslint-disable prefer-const */
export let MAX_PLURAL_ENTRIES = 10;
export let MIN_PLURAL_ENTRIES = 0;
export let RANDOMNESS_ITERATIONS = 10;
