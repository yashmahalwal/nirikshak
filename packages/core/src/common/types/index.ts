export { SchemaHelpers } from "./helpers";
export { isCustomFunction, CustomFunction } from "./custom";
export { isFakerType } from "./fakerTypes";
