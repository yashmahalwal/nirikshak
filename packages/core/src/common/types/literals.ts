/* Literals that can be provided to describe the resource */

// Basic literal type
// Ex: 1, false, null, "myString"
export type Literal = string | number | null | boolean;

// type guard: Literal
export function isLiteral(input: any): input is Literal {
    const c = typeof input;
    switch (c) {
        case "number":
        case "string":
        case "boolean":
            return true;
        default:
            return input === null;
    }
}
