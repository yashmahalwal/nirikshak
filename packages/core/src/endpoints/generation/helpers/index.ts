export { generateBodyType } from "./bodyTypeGen";
export { generateHeaderAndStatus } from "../output/headerAndStatus";
