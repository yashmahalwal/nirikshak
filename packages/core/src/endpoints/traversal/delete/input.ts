import {
    HeaderAndQueryInstance,
    generateHeaderAndQuery,
} from "../../generation/input/headerAndQuery";
import { Inputs } from "../..";
import { ResourceInstance, SchemaHelpers } from "../../..";

export interface DeleteInput {
    semantics: HeaderAndQueryInstance;
}

// Generate input from an resource
export async function generateDeleteInput(
    input: Inputs["DELETE"],
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<DeleteInput> {
    return {
        semantics: await generateHeaderAndQuery(
            input.semantics,
            resource,
            helpers
        ),
    };
}
