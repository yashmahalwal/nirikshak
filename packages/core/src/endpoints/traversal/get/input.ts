import {
    HeaderAndQueryInstance,
    generateHeaderAndQuery,
} from "../../generation/input/headerAndQuery";
import { Inputs } from "../..";
import { ResourceInstance, SchemaHelpers } from "../../..";

export interface GetInput {
    semantics: HeaderAndQueryInstance;
}

// Generate input from an resource
export async function generateGetInput(
    input: Inputs["GET"],
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<GetInput> {
    return {
        semantics: await generateHeaderAndQuery(
            input.semantics,
            resource,
            helpers
        ),
    };
}
