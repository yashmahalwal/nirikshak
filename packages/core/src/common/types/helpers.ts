import { CustomFunction } from "./custom";
import { Literal, isLiteral } from "./literals";

export interface SchemaHelpers {
    [key: string]: CustomFunction;
}

// Ultimately, a resource fields ends in Primitive type
// Ex: {age: 23}
// Ex: {address: [{zipCode: "370445"}, {zipCode: "121342"}], classes: [1,2,3,4,5]}
export type Primitives = Literal | Array<Primitives>;
export function isPrimitives(input: any): input is Primitives {
    if (Array.isArray(input)) return input.every((i) => isPrimitives(i));
    return isLiteral(input);
}
