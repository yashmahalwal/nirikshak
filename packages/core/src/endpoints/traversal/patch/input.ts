import {
    HeaderAndQueryInstance,
    generateHeaderAndQuery,
} from "../../generation/input/headerAndQuery";
import { BodyInstance, Outputs, Inputs } from "../../types";
import { ResourceInstance, SchemaHelpers } from "../../..";
import { generateBodyType } from "../..";

export interface PatchInput {
    semantics: HeaderAndQueryInstance;
    body: BodyInstance;
}

// Generate input from an resource
export async function generatePatchInput(
    caseValue: keyof Outputs["PATCH"],
    input: Inputs["PATCH"],
    resource: ResourceInstance,
    helpers: SchemaHelpers
): Promise<PatchInput> {
    if (caseValue === "DESTRUCTIVE" && !input.destructiveBody)
        throw new Error(
            `Did not find a destructive body description. Cannot make destrcutive post request`
        );
    return {
        semantics: await generateHeaderAndQuery(
            input.semantics,
            resource,
            helpers
        ),
        body: await generateBodyType(
            caseValue === "DESTRUCTIVE" ? input.destructiveBody! : input.body,
            resource,
            helpers
        ),
    };
}
