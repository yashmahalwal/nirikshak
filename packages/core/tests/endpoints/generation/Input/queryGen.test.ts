import faker from "faker";
import { Query } from "../../../../src/endpoints/types/input";
import {
    QueryInstance,
    generateQuery,
} from "../../../../src/endpoints/generation/input/queryGen";
import { SchemaHelpers } from "../../../../src/common/types/helpers";
import { Literal } from "../../../../src/common/types/literals";
import { RANDOMNESS_ITERATIONS } from "../../../../src/common/Env";

const resource = {
    id: faker.random.uuid(),
    name: {
        firstName: faker.name.firstName(),
        lastName: faker.name.lastName(),
    },
    addresses: [
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
        { zipCode: faker.address.zipCode(), city: faker.address.city() },
    ],
    recursion: {
        bases: {
            args: {
                plethora: [1, 2, 10, [[7, 5, [4]]], 3],
            },
        },
    },
};

const Helpers: SchemaHelpers = {
    word: (): Promise<string> =>
        new Promise((resolve) =>
            setTimeout(() => resolve(faker.random.word()), 10)
        ),
    async invalidHelper() {
        return { age: 10 } as any;
    },
};

const ValidQueries: {
    input: Query;
    output: () => Promise<QueryInstance>;
}[] = [
    {
        input: {
            entry1: "foo-bar",
            entry2: false,
            entry3: [
                1,
                false,
                true,
                "resource:addresses[0].zipCode",
                "faker:random.number",
                "custom:word",
            ],
        },
        async output(): Promise<QueryInstance> {
            const promiseMap: Map<
                string,
                Promise<Literal | Literal[]>
            > = new Map();

            for (const key in this.input) {
                switch (key) {
                    case "entry1":
                        promiseMap.set(key, Promise.resolve("foo-bar"));
                        break;
                    case "entry2":
                        promiseMap.set(key, Promise.resolve(false));
                        break;
                    case "entry3":
                        promiseMap.set(
                            key,
                            Promise.resolve([
                                1,
                                false,
                                true,
                                resource.addresses[0].zipCode,
                                faker.random.number(),
                                await Helpers.word(),
                            ])
                        );
                        break;
                }
            }

            const queries: QueryInstance = {};
            for (const key in this.input) {
                const res = await promiseMap.get(key)!;

                if (Array.isArray(res) && res.some((entry) => entry === null))
                    throw new Error(
                        `query: ${key}:${this.input[key]} resolved to an array containing a null value`
                    );
                else if (res === null)
                    throw new Error(
                        `query: ${key}:${this.input[key]} resolved to null`
                    );

                queries[key] = res as
                    | Exclude<Literal, null>
                    | Exclude<Literal, null>[];
            }

            return queries;
        },
    },
];

const InvalidQueries: any[] = [
    { entry1: null },
    { entry2: [true, false, "str", "strstr", null] },
    {
        entry2: [
            true,
            false,
            "str",
            "strstr",
            "resource:addresses",
            "resource:addresses",
            "resource:addresses",
        ],
    },
    { entry3: "custom:nonExistent" },
    { entry4: ["possums", "resource:lorem.ipsum"] },
    { entry5: "custom:invalidHelper" },
];

describe("Query type generation", () => {
    for (let i = 1; i <= RANDOMNESS_ITERATIONS; i++)
        ValidQueries.forEach((entry, index) => {
            test(`Valid query entry: ${index}, iteration: ${i}`, async () => {
                faker.seed(123 * i + index);
                const got = await generateQuery(entry.input, resource, Helpers);
                faker.seed(123 * i + index);
                const expected = await entry.output();
                expect(got).toEqual(expected);
            });
        });

    test.each(InvalidQueries)(`Invalid query entry: %#`, (entry) =>
        expect(
            generateQuery(entry, resource, Helpers)
        ).rejects.toMatchSnapshot()
    );
});
