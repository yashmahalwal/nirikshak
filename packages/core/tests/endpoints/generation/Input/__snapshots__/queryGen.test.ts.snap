// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Query type generation Invalid query entry: 0 1`] = `[Error: query: entry1:null resolved to null]`;

exports[`Query type generation Invalid query entry: 1 1`] = `[Error: query: entry2:true,false,str,strstr, resolved to an array containing a null value]`;

exports[`Query type generation Invalid query entry: 2 1`] = `[Error: Resource string resource:addresses resolved to non primitive value]`;

exports[`Query type generation Invalid query entry: 3 1`] = `[Error: function: custom:nonExistent, args: ]`;

exports[`Query type generation Invalid query entry: 4 1`] = `[Error: Invalid resource instance path: lorem.ipsum]`;

exports[`Query type generation Invalid query entry: 5 1`] = `[Error: Non literal type value from custom:invalidHelper]`;
